heft-plugins/heft-esbuild-plugin/src/EsbuildPlugin.ts

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import { ITerminal, Import } from '@rushstack/node-core-library';
+import type {
+  HeftConfiguration,
+  HeftSession,
+  ICompileSubstage,
+  IBuildStageContext,
+  IHeftPlugin,
+  ScopedLogger
+} from '@rushstack/heft';
+
+import { configurationFileLoader, IEsbuildConfigurationJson } from './EsbuildConfigLoader.js';
+import { BuildFailure, BuildResult } from 'esbuild';
+
+const esbuild: typeof import('esbuild') = Import.lazy('esbuild', require);
+
+const PLUGIN_NAME: string = 'EsbuildPlugin';
+
+interface IEsbuildConfigurationFileCacheEntry {
+  configurationFile: IEsbuildConfigurationJson | undefined;
+}
+
+/**
+ * @internal
+ */
+export class EsbuildPlugin implements IHeftPlugin {
+  public readonly pluginName: string = PLUGIN_NAME;
+
+  private _esbuildConfigurationFileCache: Map<string, IEsbuildConfigurationFileCacheEntry> = new Map<
+    string,
+    IEsbuildConfigurationFileCacheEntry
+  >();
+
+  public apply(heftSession: HeftSession, heftConfiguration: HeftConfiguration): void {
+    const logger: ScopedLogger = heftSession.requestScopedLogger('esbuild');
+    heftSession.hooks.build.tap(PLUGIN_NAME, (build: IBuildStageContext) => {
+      build.hooks.compile.tap(PLUGIN_NAME, (compile: ICompileSubstage) => {
+        compile.hooks.run.tapPromise(PLUGIN_NAME, async () => {
+          await new Promise<void>((resolve: () => void, reject: (error: Error) => void) => {
+            let isFirstEmit: boolean = true;
+            this._runEsbuildAsync(logger, heftConfiguration, build.properties.watchMode, () => {
+              if (isFirstEmit) {
+                isFirstEmit = false;
+                if (build.properties.watchMode) {
+                  resolve();
+                }
+              } else {
+                compile.hooks.afterRecompile.promise().catch((error) => {
+                  heftConfiguration.globalTerminal.writeErrorLine(
+                    `An error occurred in an afterRecompile hook: ${error}`
+                  );
+                });
+              }
+            })
+              .then(resolve)
+              .catch(reject);
+          });
+        });
```

Comment:
This is a little funky.

Reply:
Yeah agreed, is there a cleaner solution here?